fix `LogFFE` for FFEs not in `IsInternalRep`

ThomasBreuer opened the pull request:
resolves #3784

- remove `FFECONWAY.LogFFERhoIterate`, `FFECONWAY.DoLogFFERho`, `FFECONWAY.DoLogFFE`
- copy the code of `DLogShanks`, `DLog`, and the `LogFFE` method from `lib/Util.gi` of the StandardFF package to `lib/ffeconway.gi`, and bind the latter function to `FFECONWAY.DoLogFFE` and thus install it as a `LogFFE` method in the three relevant situations
- add some tests

(A comment in the now removed code says
> \# Adapted from the integer code by AH and Sean Gage

thus perhaps also this underlying code should be fixed.)

ThomasBreuer commented:
The hpcgap test fails.
The reason is that due to the extended test  code, the process requires more and more memory, and then the CI test gets canceled.
When I start hpcgap locally, I can reproduce the problem as follows.
```
gap> ForAll( Primes, p -> LogFFE( Z(p^2)^4, Z(p^2)^2 ) = 2 );
GC Warning: Repeated allocation of very large block (appr. size 3726905344):
	May lead to memory leak and poor performance
```
(And then I have to kill the process.)

This happens also if the StandardFF package is loaded, that is, it seems to be independent of the code change from this pull request.

fingolfin commented:
It would indeed be interesting if `DoLogModRho` could also be removed by this code (or a variation of it), depending on how it affects performance (well, and correctness). But as far as I know, no errors have been reported for that yet (resp. the function `LogMod` using it). Though I also don't know if anyone tried seriously to e.g. adapt the inputs leading to failures in `LogFFE` for `LogMod`...?

ThomasBreuer commented:
@fingolfin Your concerns about `LogMod` and `DoLogModRho` arise from the similarity of the underlying code with that of the old `LogFFE` method and `FFECONWAY.DoLogFFERho`.

`LogMod` checks whether `DoLogModRho` has failed, and falls back to another strategy in this case, thus we do not get a wrong answer, contrary to the old situation with `LogFFE`.
It might be worth to look for performance improvements for `LogMod`.

fingolfin commented:
Regarding the HPC-GAP failure, there is a bug in the kernel, I'll submit a fix momentarily.

ThomasBreuer commented:
> Regarding the HPC-GAP failure, there is a bug in the kernel, I'll submit a fix momentarily.

Great.
Then I can rebase my pull request, and the failure is expected to disappear.

fingolfin commented on hpcgap/lib/ffeconway.gi:
```diff
@@ -1196,225 +1196,93 @@ end);
-        Qp:=Qp/theta;
-    od;
-    # process failed (because z was not a primitive root)
+FFECONWAY.DoLogFFE:= function(x, base)
```
This code is in no way specific to FFECONWAY, is it? I mean, StandardFF was using it, and looking at it, it seems to be that it might in fact work for `x` and `base` elements of arbitrary multiplicative groups, right?

So perhaps it would make sense to move it elsewhere, and out of this record? And possibly drop the `FFE` from the name?

It could go into a new file, `dlog.gi` or whatever. Or an existing file (that would then also remove the need of maintaining two copies of this code for GAP vs. HPC-GAP.

ThomasBreuer commented:
O.k., I have moved the code to other files; this is clearly an improvement concerning the code duplication in `hpcgap/lib`..

Besides that, my idea is that first we get this code into GAP, such that the `LogFFE` bug is gone, and then StandardFF can be changed to not define the names `DLog`, `DLogShanks` anymore.
Afterwards we can improve the GAP library part (document the functions in the GAP Reference Manual, perhaps change the code, the function names, the order of arguments).

fingolfin reviewed (approved):
Looks good to me, thanks!

ThomasBreuer commented:
@luebeck As I had sketched, my idea was to do the code move in several steps:

The first step is to fix the bug in the GAP library by using code copied from StandardFF, such that the current StandardFF can still be loaded in the master branch version of GAP. 
The second step is that a new version of StandardFF takes care of a possible presence of the relevant functions in GAP.
The third step is to adjust the GAP library to the new StandardFF.

Since `DLog` and `DLogShanks` are not operations but global function, creating variables with these names in GAP will make it impossible to load StandardFF in the master branch version of GAP. (Note that the current AtlasRep suggests loading (the current version of) StandardFF.)
After the abovementioned second step, users of the master branch version of GAP can upgrade their StandardFF.
(I do not see how we could proceed without such a forced upgrade.)

fingolfin commented:
We could move forward with using `DLog` and `DLogShanks` here if there first was a new StandardFF release which defines these only conditionally. Coincidentally, I already [requested](https://github.com/frankluebeck/StandardFF/issues/5) a new StandardFF release :-)

ThomasBreuer commented:
@luebeck Thanks for the StandardFF release.
- I have updated this pull request such that the functions `DLog` and `DLogShanks` get defined in the GAP library.
  Thus users of the dev version of GAP will have to upgrade their StandardFF installation as soon as this pull request gets merged.
- And I have added the original documentation of `DLog` from StandardFF to the GAP Reference Manual.
  It is now in Section "Primitive Roots and Discrete Logarithms" (in Chapter "Number Theory"), which contains also `OrderMod` and `LogMod`. This is not an optimal place, but also the neighborhood of `LogFFE` (in Section "Operations for Finite Field Elements", Chapter "Finite Fields") does not fit well.
  (Isn't `DLog` in fact more general than stated in the documentation, in the sense that the argument `x` is not required to be a power of `base`, and that `fail` is returned if not?)
- Concerning the idea to use `DoDLog` also in the two `LogFFE` methods "for two elements in Z/pZ (ModulusRep)" and "for two FFEs (in a prime field)", I have not changed these methods because I am not sure how to proceed. Some experiments will be necessary in order to find a good heuristic. Moreover, both methods call `LogMod`, thus it would perhaps be more natural to change `LogMod`.

frankluebeck reviewed (approved):
Thanks, looks good now.
